src/meta/KVBasedGflagsManager.cpp

```diff
@@ -0,0 +1,141 @@
+ */
+
+#include "meta/KVBasedGflagsManager.h"
+#include "meta/processors/Common.h"
+#include "meta/MetaServiceUtils.h"
+
+DECLARE_int32(load_config_interval_secs);
+
+namespace nebula {
+namespace meta {
+
+KVBasedGflagsManager::KVBasedGflagsManager(kvstore::KVStore* kv) {
+    kvstore_ = dynamic_cast<kvstore::NebulaStore*>(kv);
+    CHECK_NOTNULL(kvstore_);
+}
+
+KVBasedGflagsManager::~KVBasedGflagsManager() {
+    kvstore_ = nullptr;
+}
+
+Status KVBasedGflagsManager::init() {
+    getGflagsModule();
+    declareGflags();
+    return registerGflags();
+}
+
+folly::Future<StatusOr<bool>>
+KVBasedGflagsManager::setConfig(const cpp2::ConfigModule& module, const std::string& name,
+                                const cpp2::ConfigType& type, const VariantType &value) {
+    UNUSED(module); UNUSED(name); UNUSED(type); UNUSED(value);
+    LOG(FATAL) << "Unimplement!";
+    return Status::NotSupported();
+}
+
+folly::Future<StatusOr<std::vector<cpp2::ConfigItem>>>
+KVBasedGflagsManager::getConfig(const cpp2::ConfigModule& module, const std::string& name) {
+    UNUSED(module); UNUSED(name);
+    LOG(FATAL) << "Unimplement!";
+    return Status::NotSupported();
+}
+
+folly::Future<StatusOr<std::vector<cpp2::ConfigItem>>>
+KVBasedGflagsManager::listConfigs(const cpp2::ConfigModule& module) {
+    UNUSED(module);
+    LOG(FATAL) << "Unimplement!";
+    return Status::NotSupported();
+}
+
+folly::Future<StatusOr<bool>>
+KVBasedGflagsManager::registerConfig(const cpp2::ConfigModule& module, const std::string& name,
+                                     const cpp2::ConfigType& type, const cpp2::ConfigMode& mode,
+                                     const std::string& value) {
+    UNUSED(module); UNUSED(name); UNUSED(type); UNUSED(mode); UNUSED(value);
+    LOG(FATAL) << "Unimplement!";
+    return Status::NotSupported();
+}
+
+void KVBasedGflagsManager::getGflagsModule() {
+    // get current process according to gflags pid_file
+    gflags::CommandLineFlagInfo pid;
+    if (gflags::GetCommandLineFlagInfo("pid_file", &pid)) {
+        auto defaultPid = pid.default_value;
+        if (defaultPid.find("nebula-graphd") != std::string::npos) {
+            module_ = cpp2::ConfigModule::GRAPH;
+        } else if (defaultPid.find("nebula-storaged") != std::string::npos) {
+            module_ = cpp2::ConfigModule::STORAGE;
+        } else if (defaultPid.find("nebula-metad") != std::string::npos) {
+            module_ = cpp2::ConfigModule::META;
+        } else {
+            LOG(ERROR) << "Should not reach here";
+        }
+    } else {
+        LOG(ERROR) << "Should not reach here";
+    }
+}
+
+Status KVBasedGflagsManager::registerGflags() {
+    // based on kv, so we will retry until register config successfully
+    bool succeeded = false;
+    while (!succeeded) {
+        std::vector<kvstore::KV> data;
+        folly::SharedMutex::WriteHolder wHolder(LockUtils::configLock());
```

Comment:
why need write lock?

Reply:
Maybe you need to refer to PR #622.

Reply:
like you say this function just call once from init in main, and this will block the main thread, then the whole process just one thread, so no need lock?

Reply:
Yeah, just in case.